src/gluonnlp/models/t5.py

```diff
@@ -0,0 +1,849 @@
+                T5Block(
+                    d_model=d_model, 
+                    d_kv=d_kv, 
+                    d_ff=d_ff, 
+                    is_decoder=is_decoder, 
+                    num_heads=num_heads, 
+                    dropout_prob=dropout_prob, 
+                    layer_norm_eps=layer_norm_eps, 
+                    activation=activation, 
+                    init_factor=init_factor, 
+                    layout=layout, 
+                    dtype=dtype
+                )
+            )
+        self.final_layer_norm = T5LayerNorm(
+            d_model=d_model, 
+            epsilon=layer_norm_eps, 
+            init_factor=init_factor, 
+            dtype=dtype
+        )
+        self.dropout = nn.Dropout(dropout_prob)
+
+    @property
+    def layout(self): 
+        return self._layout
+
+    @_assert_decoder_method
+    @property
+    def state_batch_axis(self): 
+        return list(layer.state_batch_axis for layer in self.layers)
+
+    @_assert_decoder_method
+    def init_states(self, batch_size, ctx, dtype='float32'): 
+        return list(layer.init_states(batch_size, ctx, dtype) for layer in self.layers)
+
+    @_assert_decoder_method
+    def incremental_decode(
+        hidden_states, 
+        past_key_value, 
+        mem_states, 
+        mem_valid_length
+    ): 
+        raise NotImplementedError
+
+    def _get_query_key_shape(self, hidden_states, mem_states, past_key_value): 
+        # NT: (B, L_seq, inner_dim); TN: (L_seq, B, inner_dim)
+        index = 1 if self.layout == 'NT' else 0
+        query_length = hidden_states.shape[index]
+        if past_key_value is not None: 
+            # for incremental decoding only, where past key and value are of shape
+            # NT(NTK): (B, L_seq, num_heads, n_kv); TN(TNK): (L_seq, B, num_heads, n_kv)
+            query_length += past_key_value[0].shape[index]
+        key_length = query_length if mem_states is None else mem_states.shape[index]
+        return query_length, key_length
+
+    @staticmethod
+    def _get_relative_position(query_length, key_length): 
+        # relative_position = mem_i - query_j
+        query_position = np.arange(query_length, dtype=np.int32)[:, None]
+        mem_position = np.arange(key_length, dtype=np.int32)[None, :]
+        relative_position = mem_position - query_position
+        return relative_position
+
+    def forward(
+        self, 
+        hidden_states, 
+        valid_length, 
+        mem_states=None, 
+        mem_valid_length=None
+    ): 
+        # 1. relative position embeddings and attention masks
+        position_embeddings = self.relative_position_encoder(
+            T5Stack._get_relative_position(*self._get_query_key_shape(hidden_states, None, None))
+        )
+        if self._is_decoder: 
+            mem_position_embeddings = np.zeros(
+                (1, self._num_heads, *self._get_query_key_shape(hidden_states, mem_states, None)), 
+                dtype=self._dtype
+            )
```

Comment:
Try to use zeros_like.

Reply:
You may try to first use `arange_like` to get the two ndarrays, one with shape (query_length,), and another with shape (mem_length,). After that, you can use these two arrays to get `(query_length, mem_length)` and later get `(1, num_heads, query_length, mem_length)`.